README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# smallfactor - store factors in bytes and bits rather than an `integer`

<!-- badges: start -->

![](https://img.shields.io/badge/cool-useless-green.svg)
[![R-CMD-check](https://github.com/coolbutuseless/smallfactor/workflows/R-CMD-check/badge.svg)](https://github.com/coolbutuseless/smallfactor/actions)
<!-- badges: end -->

`smallfactor` is an experiment to see what trade-offs there might be for
storing a factor in bytes or bits instead of an integer.

An R integer is 4-bytes, meaning it could hold a factor with two billion
levels, which seems like overkill.

`bytefactor` targets factors with up to 256 levels.

`bitfactor` will choose the fewest bits to store a factor for factors
with 2 up to 32768 (2^15) levels.

## What’s in the box

-   `bytefactor()` store factor values in raw bytes rather than an
    integer.
-   `bitfactor()` store factor values as bits within integers - with
    multiple factors stored in a single integer value. E.g. a factor
    with 4 levels could be stored as just 2 bits of information, meaning
    16 values could be stored in a single R integer (16 \* 2 bits = 32
    bits = 4 bytes)

## Limitations

It seems possible that you could write more methods to make these
smaller factors behave similar to regular R `factor` objects, but this
package is not attempting this (yet).

Internally, there’s currently a lot of transferring back-and-forth
between these small factors and the standard R factor in order to make
use of the printing and subsetting capabilities of the R factor
implementation. Much of this back-and-forth could be avoided if effort
was expended to do so.

Note: `bitfactor` uses only 31 bits of a 32-bit integer in order to
avoid issues around `NA_integer_` representation. This means, for
example, that an integer can only hold 15 x 2-bit values. In practice
the user is never expected to notice this or care about it.

## Installation

You can install from
[GitHub](https://github.com/coolbutuseless/smallfactor) with:

``` r
# install.package('remotes')
remotes::install_github('coolbutuseless/smallfactor')
```

## `bytefactor`

``` r
small <- bytefactor(c('a', 'b', 'c', 'a', 'd'))
small
```

    #> [bytefactor]
    #> [1] a b c a d
    #> Levels: a b c d

``` r
small[1:3]
```

    #> [bytefactor]
    #> [1] a b c
    #> Levels: a b c d

## `bitfactor`

`bitfactor()` will choose an appropriate number of bits to store the
given number of levels.

In the following example, there are 4 levels, so `bitfactor()` chooses
to store each value in the factor in 2 bits.

``` r
tiny <- bitfactor(c('a', 'b', 'c', 'a', 'd'))
tiny
```

    #> [bitfactor] 5 values @ ~2 bits/value = 1 integer(s)
    #> [1] a b c a d
    #> Levels: a b c d

``` r
tiny[1:3]
```

    #> [bitfactor] 3 values @ ~2 bits/value = 1 integer(s)
    #> [1] a b c
    #> Levels: a b c d

In this next example, there are 100 levels in the factor, so 7 bits are
needed to fully store all the levels

``` r
tiny <- bitfactor(sample(100, 10), levels=1:100)
tiny
```

    #> [bitfactor] 10 values @ ~7 bits/value = 3 integer(s)
    #>  [1] 68 39 1  34 87 43 14 82 59 51
    #> 100 Levels: 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 ... 100

``` r
tiny[4:6]
```

    #> [bitfactor] 3 values @ ~7 bits/value = 1 integer(s)
    #> [1] 34 87 43
    #> 100 Levels: 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 ... 100

## Storing some DNA in `bytefactor` and `bitfactor` objects.

|                | character | factor | bytefactor | bitfactor |
|----------------|-----------|--------|------------|-----------|
| bits/value     | 64        | 32     | 8          | 2         |
| total size     | 8 MB      | 4 MB   | 1 MB       | 270 kB    |
| size reduction |           | 2x     | 8x         | 30x       |

``` r
library(smallfactor)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Generate some random DNA
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
character_vector_dna <- sample(c('A', 'T', 'G', 'C'), 1e6, replace = TRUE)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create a `factor` and a `smallfactor` using the same basic syntax
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
integer_factor <- factor    (character_vector_dna, levels = c('A', 'T', 'G', 'C'))
byte_factor    <- bytefactor(character_vector_dna, levels = c('A', 'T', 'G', 'C'))
bit_factor     <- bitfactor (character_vector_dna, levels = c('A', 'T', 'G', 'C'))

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# `smallfactor` is approx 1/4 the size of the regular factor
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lobstr::obj_size(character_vector_dna)
```

    #> 8,000,272 B

``` r
lobstr::obj_size(integer_factor)
```

    #> 4,000,688 B

``` r
lobstr::obj_size(byte_factor)
```

    #> 1,000,696 B

``` r
lobstr::obj_size(bit_factor)
```

    #> 267,704 B

## Similar projects

-   [{lofi}](https://github.com/coolbutuseless/lofi)

## Acknowledgements

-   R Core for developing and maintaining the language.
-   CRAN maintainers, for patiently shepherding packages onto CRAN and
    maintaining the repository


NEWS.md
# smallfactor 0.1.1 2021-09-02

* Rename `smallfactor()` to `bytefactor()`
* Add `bitfactor()`
* Add tests

* Initial release

# smallfactor 0.1.0  2021-08-27

* Initial release


tests/testthat.R
library(testthat)
library(smallfactor)

test_check("smallfactor")


tests/testthat/test-pack.R




test_that("pack structure definitions", {

  expect_equal(base[[1]], 1L )
  expect_equal(base[[2]], 1L + 2L)
  expect_equal(base[[3]], 1L + 2L + 4L)


  expect_equal(
    shifts[[2]],
    seq.int(0, 31 - 2, 2)
  )

  expect_equal(
    shifts[[3]],
    seq.int(0, 31 - 3, 3)
  )

  expect_equal(
    shifts[[4]],
    seq.int(0, 31 -4 , 4)
  )

  expect_equal(
    shifts[[5]],
    seq.int(0, 31 - 5, 5)
  )


  expect_equal(
    masks[[2]],
    bitwShiftL(3,  seq(0, 31-2, 2))
  )

  expect_equal(
    masks[[3]],
    bitwShiftL(7,  seq.int(0, 31-3, 3))
  )

})



test_that("single int works", {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  int <- pack_into_single_int(c(1, 2, 3, 4), nbits = 7)

  expect_equal(
    int,
    sum(
      bitwShiftL(1,  0),
      bitwShiftL(2,  7),
      bitwShiftL(3, 14),
      bitwShiftL(4, 21)
    )
  )

  res <- unpack_from_single_int(int, nbits = 7)
  expect_equal(res, c(1, 2, 3, 4))



  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  int <- pack_into_single_int(c(1, 2), nbits = 7)

  expect_equal(
    int,
    sum(
      bitwShiftL(1,  0),
      bitwShiftL(2,  7),
      bitwShiftL(1, 14),
      bitwShiftL(2, 21)
    )
  )

  res <- unpack_from_single_int(int, nbits = 7)
  expect_equal(res, c(1, 2, 1, 2))



  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  int <- pack_into_single_int(c(1, 2), nbits = 15)

  expect_equal(
    int,
    sum(
      bitwShiftL(1,  0),
      bitwShiftL(2, 15)
    )
  )

  res <- unpack_from_single_int(int, nbits = 15)
  expect_equal(res, c(1, 2))
})


test_that("multiple ints works", {
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ints <- pack_ints(c(1, 2, 3, 4, 5, 6, 7, 8), nbits = 8)
  res  <- unpack_ints(ints, nbits = 8)
  expect_equal(res, c(1, 2, 3, 4, 5, 6, 7, 8))
})


test_that("multiple ints stress test", {

  set.seed(2021)
  for (i in seq(200)) {
    N     <- sample(50, 1)
    nbits <- sample(1:15, 1)
    orig  <- sample(0:(2^nbits-1), N, replace = TRUE)
    ints  <- pack_ints(orig, nbits = nbits)
    res   <- unpack_ints(ints, nbits = nbits)
    expect_identical(res, orig)
  }

})










tests/testthat/test-bitfactor.R




test_that("bitfactor works", {


  x <- bitfactor(1:10)
  expect_true(inherits(x, 'bitfactor'))

  expect_length(x, 10)

  expect_equal(as.integer(x), 1:10)


  expect_equal(
    x[1:5],
    bitfactor(1:5, levels = as.character(1:10))
  )

  capture_output(
  expect_true(
    {print(x); as.character(x); TRUE}
  )
  )

  x0 <- factor(1:10)
  x1 <- as.bitfactor(x0)
  expect_equal(
    x1,
    x
  )

  expect_equal(
    as.double(x),
    1:10
  )

})


R/pack.R


base <- lapply(
  1:15,
  function(nbits) {
    as.integer(sum(2 ^ seq.int(0, nbits-1)))
  }
)

shifts <- lapply(1:15, function(nbits) {
  (seq.int(31 %/% nbits) - 1L) * nbits
})

create_masks <- function(n) { # nocov start
  n      <- as.integer(n)
  b      <- base[[n]]
  ss     <- shifts[[n]]
  vapply(
    ss,
    function(shift) {
      bitwShiftL(b, shift)
    },
    integer(1)
  )
} # nocov end

masks <- lapply(1:15, create_masks)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Pack integer values into a single integer
#'
#' @param small_ints vector of small integer values. these are recycle if necessary
#' @param nbits number of bits we want to pack these into
#'
#' @return single integer value
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pack_into_single_int <- function(small_ints, nbits) {
  # ToDo: assert that length of small_ints matches expectations of 'nbits'
  # i.e. if nbits=16, then small_ints can only have a maximum of 2 values
  maxlen <- floor(31/nbits)
  stopifnot(length(small_ints) <= maxlen)

  shifted_small_ints <-
    bitwShiftL(
      bitwAnd(small_ints, base[[nbits]]),  # ensure bits are within range
      shifts[[nbits]]
    )

  sum(shifted_small_ints)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Unpacka an integer value into a vector of small integers of 'nbits' each
#'
#' @param int single integer value
#' @inheritParams pack_into_single_int
#'
#' @return vector of small integers
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
unpack_from_single_int <- function(int, nbits) {
  shifted_ints <- bitwAnd(int, masks[[nbits]])
  bitwShiftR(shifted_ints, shifts[[nbits]])
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Pack an unbounded vector of small integers into 'nbit' values inside a a vector of 32bit integers
#'
#' @inheritParams pack_into_single_int
#'
#' @return vector of integer values
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pack_ints <- function(small_ints, nbits) {
  N <- 31L %/% nbits
  chunks <- chunk(small_ints, N)
  res <- vapply(chunks, pack_into_single_int, integer(1), nbits = nbits)

  attr(res, 'N') <- length(small_ints)
  res
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Unpack a 32-bit integer values into a vector of small integers of 'nbits' each
#'
#' @param ints integer vector
#' @inheritParams unpack_from_single_int
#'
#' @return vector of small integers
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
unpack_ints <- function(ints, nbits) {
  res <- unlist(lapply(ints, unpack_from_single_int, nbits = nbits))

  N <- attr(ints, 'N', exact = TRUE)
  if (!is.null(N)) {
    res <- res[seq.int(N)]
  }

  res
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Split a vector in chunks of size 'n'
#'
#' @param x vector
#' @param n chunk size
#'
#' @return list of vectors
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
chunk <- function(x, n) {
  N <- length(x)
  S <- seq.int(from=1L, to=N, by=n)
  mapply(
    function(a, b) {x[a:b]},
    S,
    pmin(S + (n - 1L), N),
    SIMPLIFY = FALSE
  )
}


R/bytefactor.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Subset a bytefactor
#'
#' @param x bytefactor
#' @param ... ignored
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
'[.bytefactor' <- function(x, ...) {
  y <- NextMethod("[[")
  attr(y, "levels") <- attr(x, "levels")
  class(y) <- oldClass(x)
  y
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Include a bytefactor in a data.frame object
#'
#' Code based on \code{as.data.frame.numeric}
#'
#' @param x,row.names,optional,...,nm see documentation for \code{as.data.frame.faactor()}
#'
#' @return data.frame
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.data.frame.bytefactor <- function (x, row.names = NULL, optional = FALSE, ..., nm = deparse1(substitute(x))) { # nocov start
  force(nm)
  nrows <- length(x)
  if (!(is.null(row.names) || (is.character(row.names) && length(row.names) ==
                               nrows))) {
    warning(gettextf("'row.names' is not a character vector of length %d -- omitting it. Will be an error!",
                     nrows), domain = NA)
    row.names <- NULL
  }
  if (is.null(row.names)) {
    if (nrows == 0L)
      row.names <- character()
    else if (length(row.names <- names(x)) != nrows || anyDuplicated(row.names))
      row.names <- .set_row_names(nrows)
  }
  if (!is.null(names(x)))
    names(x) <- NULL
  value <- list(x)
  if (!optional)
    names(value) <- nm
  structure(value, row.names = row.names, class = "data.frame")
} # nocov end


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert an ordinary factor into a bytefactor
#'
#' @param x factor object
#'
#' @return bytefactor object if number of levels < 256, otherwise error
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.bytefactor <- function(x) {
  stopifnot(`only factor objects supported for now` = is.factor(x))
  stopifnot(length(levels(x)) < 256)

  bytefactor(x, levels = levels(x))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a bytefactor into a regular factor
#'
#' @param x bytefactor object
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.factor.bytefactor <- function(x) {
  factor(as.integer(unclass(x)), levels = seq.int(length(levels(x))), labels = levels(x))
  # factor(as.integer(unclass(x)), levels = levels(x))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Other representations of a small factor
#'
#' @param x bytefactor object
#' @param ... other arguments passed to \code{as.character.factor()}
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.character.bytefactor <- function(x, ...) {
  as.character(as.factor.bytefactor(x), ...)
}

#' @rdname as.character.bytefactor
#' @export
as.integer.bytefactor <- function(x, ...) {
  as.integer(as.factor.bytefactor(x), ...)
}

#' @rdname as.character.bytefactor
#' @export
as.double.bytefactor <- function(x, ...) {
  as.double(as.factor.bytefactor(x), ...)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Print a bytefactor
#'
#' @param x bytefactor object
#' @param ... other arguments passed to \code{print.factor()}
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
print.bytefactor <- function(x, ...) {
  cat("[bytefactor]\n")
  print(as.factor.bytefactor(x), ...)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create a bytefactor i.e. a factor backed by a raw byte vector, not an integer
#'
#' @param x object to turn into a factor
#' @param levels the levels of the factor. Default: NULL means to take the
#'        level names from the given object
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bytefactor <- function(x, levels = NULL) {

  stopifnot(length(unique(x)) < 256)

  if (is.null(levels)) {
    y      <- unique(x)
    ind    <- order(y)
    levels <- unique(as.character(y)[ind])
  }

  x <- as.character(x)
  f <- as.raw(match(x, levels))
  attr(f, 'levels') <- levels

  class(f) <- c('bytefactor')
  f
}



R/smallfactor.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Subset a smallfactor
#'
#' @param x smallfactor
#' @param ... ignored
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
'[.smallfactor' <- function(x, ...) {
  y <- NextMethod("[[")
  attr(y, "levels") <- attr(x, "levels")
  class(y) <- oldClass(x)
  y
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Include a smallfactor in a data.frame object
#'
#' @param x,row.names,optional,...,nm see documentation for \code{as.data.frame.faactor()}
#'
#' @return data.frame
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.data.frame.smallfactor <- function (x, row.names = NULL, optional = FALSE, ..., nm = deparse1(substitute(x))) {
  force(nm)
  nrows <- length(x)
  if (!(is.null(row.names) || (is.character(row.names) && length(row.names) ==
                               nrows))) {
    warning(gettextf("'row.names' is not a character vector of length %d -- omitting it. Will be an error!",
                     nrows), domain = NA)
    row.names <- NULL
  }
  if (is.null(row.names)) {
    if (nrows == 0L)
      row.names <- character()
    else if (length(row.names <- names(x)) != nrows || anyDuplicated(row.names))
      row.names <- .set_row_names(nrows)
  }
  if (!is.null(names(x)))
    names(x) <- NULL
  value <- list(x)
  if (!optional)
    names(value) <- nm
  structure(value, row.names = row.names, class = "data.frame")
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert an ordinary factor into a smallfactor
#'
#' @param x factor object
#'
#' @return smallfactor object if number of levels < 256, otherwise error
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.smallfactor <- function(x) {
  stopifnot(`only factor objects supported for now` = is.factor(x))
  stopifnot(length(levels(x)) < 256)

  smallfactor(x, levels = levels(x))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a smallfactor into a regular factor
#'
#' @param x smallfactor object
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.factor.smallfactor <- function(x) {
  factor(as.integer(x), levels = seq.int(length(levels(x))), labels = levels(x))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Character representation of a small factor
#'
#' @param x smallfactor object
#' @param ... other arguments passed to \code{as.character.factor()}
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.character.smallfactor <- function(x, ...) {
  as.character(as.factor.smallfactor(x), ...)
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Print a smallfactor
#'
#' @param x smallfactor object
#' @param ... other arguments passed to \code{print.factor()}
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
print.smallfactor <- function(x, ...) {
  cat("[smallfactor]\n")
  print(as.factor.smallfactor(x), ...)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create a smallfactor i.e. a factor backed by a raw byte vector, not an integer
#'
#' @param x object to turn into a factor
#' @param levels the levels of the factor. Default: NULL means to take the
#'        level names from the given object
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
smallfactor <- function(x, levels = NULL) {

  stopifnot(length(unique(x)) < 256)

  if (is.null(levels)) {
    y      <- unique(x)
    ind    <- order(y)
    levels <- unique(as.character(y)[ind])
  }

  x <- as.character(x)
  f <- as.raw(match(x, levels))
  attr(f, 'levels') <- levels

  class(f) <- c('smallfactor')
  f
}



R/bitfactor.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Subset a bitfactor
#'
#' ToDo: This currently converts the entire bitfactor into a regular R
#' factor, subsets that R factor, then converts the subset back into a
#' bitfactor.   This is horribly inefficient, and should really just extract
#' the values being asked for.
#'
#' @param x bitfactor
#' @param y subset spedification
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"[.bitfactor" <- function(x, y) {
  as.bitfactor(as.factor.bitfactor(x)[y])
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a bitfactor to a factor
#'
#' @param x bitfactor
#' @param ... ignored
#'
#' @return standard R factor
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.factor.bitfactor <- function(x, ...) {
  nbits  <- attr(x,  'nbits', exact = TRUE)
  N      <- attr(x,      'N', exact = TRUE)
  levels <- attr(x, 'levels', exact = TRUE)

  small_ints <- unpack_ints(x, nbits = nbits) + 1L
  factor(levels[small_ints], levels = levels)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' convert bitfactor to integer, numeric and character
#'
#' @inheritParams as.factor.bitfactor
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.integer.bitfactor <- function(x, ...) {
  as.integer(as.factor.bitfactor(x))
}

#' @rdname as.integer.bitfactor
#' @export
as.double.bitfactor <- function(x, ...) {
  as.double(as.factor.bitfactor(x))
}

#' @rdname as.integer.bitfactor
#' @export
as.character.bitfactor <- function(x, ...) {
  as.character(as.factor.bitfactor(x))
}

#' @rdname as.integer.bitfactor
#' @export
length.bitfactor <- function(x, ...) {
  attr(x, 'N', exact = TRUE)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a standard factor to a bitfactor
#'
#' @param x factor object
#'
#' @return bitfactor object
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
as.bitfactor <- function(x) {
  stopifnot(`only conversion from factors currently supported` = is.factor(x))

  bitfactor(as.character(x), levels(x))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Print a bitfactor
#'
#' @inheritParams as.factor.bitfactor
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
print.bitfactor <- function(x, ...) {
  nbits  <- attr(x,  'nbits', exact = TRUE)
  N      <- attr(x,      'N', exact = TRUE)
  cat("[bitfactor] ", N, " values @ ~", nbits,
      " bits/value = ", length(unclass(x)) , " integer(s)\n", sep = "")
  print(as.factor.bitfactor(x))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create a bitfactor
#'
#' @param x something coercile into a character vector
#' @param levels named levels.  This should probably just come from the
#'        values on 'x'.
#'
#' @return bitfactor object
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bitfactor <- function(x, levels = NULL) {

  # 15bits is the maximum sensible number of levels to store in a bit factor.
  # Any larger than 15 bits, and you can only fit 1 value per integer,
  # in which case you should just use a normal R factor.
  stopifnot(length(unique(x)) < 32768) # 2^15

  if (is.null(levels)) {
    y      <- unique(x)
    ind    <- order(y)
    levels <- unique(as.character(y)[ind])
  }

  x <- as.character(x)
  f <- match(x, levels) - 1L   # start at 0, rather than R's 1-indexing

  nbits <- as.integer(ceiling(log2(length(levels))))

  ints <- pack_ints(f, nbits)

  attr(ints, 'levels') <- levels
  attr(ints, 'nbits' ) <- nbits

  class(ints) <- c('bitfactor')
  ints
}



if (FALSE) {
  ints   <- c(1, 69, 255, 3, 10, 11, 12, 199, 257)

  z1 <- pack_ints(ints, nbits = 8)
  unpack_ints(z1, nbits = 8)

  orig <- sample(letters, 120, replace = TRUE)
  orig
  x <- bitfactor(orig, levels = letters)

}


